nimare/parcellate.py

```diff
@@ -0,0 +1,223 @@
+            Dataset to analyze.
+        """
+        self.dataset = dataset
+        self.masker = self.masker or dataset.masker
+
+        # Loop through voxels in target_mask, selecting studies for each and running MACMs (no MCC)
+        target_ijk = np.vstack(np.where(self.target_mask.get_fdata()))
+        target_xyz = vox2mm(target_ijk, self.masker.mask_img.affine)
+        n_target_voxels = target_xyz.shape[1]
+        n_mask_voxels = self.masker.transform(self.masker.mask_img).shape[0]
+
+        n_filters = len(getattr(self, self.filter_type))
+        labels = np.zeros((n_filters, len(self.n_clusters), n_target_voxels), dtype=int)
+        kwargs = {"r": None, "n": None}
+
+        # Use a memmapped 2D array
+        start_time = datetime.datetime.now().strftime("%Y%m%dT%H%M%S")
+        dataset_dir = _get_dataset_dir("temporary_files", data_dir=None)
+        _, memmap_filename = mkstemp(
+            prefix=self.__class__.__name__,
+            suffix=start_time,
+            dir=dataset_dir,
+        )
+        data = np.memmap(
+            memmap_filename,
+            dtype=float,
+            mode="w+",
+            shape=(n_target_voxels, n_mask_voxels),
+        )
+
+        for i_filter in range(n_filters):
+            kwargs[self.filter_type] = getattr(self, self.filter_type)[i_filter]
+            for j_coord in n_target_voxels:
+                xyz = target_xyz[:, j_coord]
+                macm_ids = dataset.get_studies_by_coordinate(xyz, **kwargs)
+                coord_dset = dataset.slice(macm_ids)
+
+                # This seems like a somewhat inelegant solution
+                # Check if the meta method is a pairwise estimator
+                if "dataset2" in inspect.getfullargspec(self.meta_estimator.fit).args:
+                    unselected_ids = sorted(list(set(dataset.ids) - set(macm_ids)))
+                    unselected_dset = dataset.slice(unselected_ids)
+                    self.meta_estimator.fit(coord_dset, unselected_dset)
+                else:
+                    self.meta_estimator.fit(coord_dset)
+
+                data[j_coord, :] = self.meta_estimator.results.get_map(
+                    self.target_image,
+                    return_type="array",
+                )  # data is overwritten across filters
+
+            # Perform clustering
+            for j_cluster, cluster_count in enumerate(self.n_clusters):
+                kmeans = KMeans(
+                    n_clusters=cluster_count,
+                    init="k-means++",
+                    n_init=10,
+                    random_state=0,
+                    algorithm="elkan",
+                ).fit(data)
+                labels[i_filter, j_cluster, :] = kmeans.labels_
+
+        # Clean up MACM data memmap
+        LGR.info(f"Removing temporary file: {memmap_filename}")
+        os.remove(memmap_filename)
+
+        images = {"labels": labels}
+        return images
+
+    def _filter_selection(self):
```

Comment:
Chase 2020:
>We implemented a two-step procedure that involved a decision on those filter sizes to be included in the final analysis and subsequently a decision on the optimal cluster solution. In the first step, we examined the consistency of the cluster assignment for the individual voxels across the cluster solutions of the co-occurrence maps performed at different filter sizes. We selected a filter range with the lowest number of deviants, that is, number of voxels that were assigned differently compared with the solution from the majority of filters. In other words, we identified those filter sizes which produced solutions most similar to the consensus-solution across all filter sizes. For example, the proportion of deviants for the second parcellation is illustrated in Figure S1; this shows the borders of the filter range to be used for subsequent steps was based on the Z-scores of the number of deviants.

I interpret this to mean:

1. Derive mode array of label assignments for each cluster count across filter sizes.
    - I assume this means mode of each voxel determined independently, rather than mode of full set of assignments.
    - What if label numbers don't match? E.g., label 1 in filter size 1 is most similar to label 2 in filter size 2.
    - I assume we should do some kind of synchronization, unless there's some inherent order to KMeans labels?
2. Count number of voxels that don't match mode for each filter size.
3. Calculate proportion of deviants in each cluster solution and filter size.
3. Calculate weighted z-score for each filter size (across cluster solutions) somehow?
    - What is it weighted by?
4. Select range of filter sizes with lowest z-scores.
    - How? Is there some kind of threshold? Figure S1 grabs range with z-scores < -0.5. No clue if that's a meaningful threshold or something like 2 standard deviations from avg z-score or what.
    - What if there are multiple dips? Does amplitude (z-scores of filters below threshold) or width (number of filters below threshold) matter more?